server/src/main/java/org/opensearch/gateway/GatewayAllocator.java

```diff
@@ -188,12 +199,26 @@ protected static void innerAllocatedUnassigned(
     public AllocateUnassignedDecision explainUnassignedShardAllocation(ShardRouting unassignedShard, RoutingAllocation routingAllocation) {
         assert unassignedShard.unassigned();
         assert routingAllocation.debugDecision();
-        if (unassignedShard.primary()) {
-            assert primaryShardAllocator != null;
-            return primaryShardAllocator.makeAllocationDecision(unassignedShard, routingAllocation, logger);
+        if (this.batchMode) {
+            if (getBatchId(unassignedShard, unassignedShard.primary()) == null) {
+                createAndUpdateBatches(routingAllocation, unassignedShard.primary());
+            }
+            assert getBatchId(unassignedShard, unassignedShard.primary()) != null;
+            if (unassignedShard.primary()) {
+                assert primaryBatchShardAllocator != null;
+                return primaryBatchShardAllocator.makeAllocationDecision(unassignedShard, routingAllocation, logger);
```

Comment:
Let's finalize on the naming once and for all

@khushbr  - we want to ahead with primary**ShardBatch**Allocator right ?

Reply:
Yes, let us keep the naming consistent. @Gaurav614 is planning to rename this to `primaryShardBatchAllocator`